src/lib/operators/table_scan/is_null_table_scan_impl.cpp

```diff
@@ -13,84 +13,77 @@
 
 namespace opossum {
 
-IsNullTableScanImpl::IsNullTableScanImpl(const std::shared_ptr<const Table>& in_table, const ColumnID base_column_id,
+IsNullTableScanImpl::IsNullTableScanImpl(const std::shared_ptr<const Table>& in_table, const ColumnID column_id,
                                          const PredicateCondition& predicate_condition)
-    : BaseSingleColumnTableScanImpl{in_table, base_column_id, predicate_condition} {
+    : AbstractSingleColumnTableScanImpl{in_table, column_id, predicate_condition} {
   DebugAssert(predicate_condition == PredicateCondition::IsNull || predicate_condition == PredicateCondition::IsNotNull,
               "Invalid PredicateCondition");
 }
 
 std::string IsNullTableScanImpl::description() const { return "IsNullScan"; }
 
-void IsNullTableScanImpl::handle_segment(const ReferenceSegment& base_segment,
-                                         std::shared_ptr<SegmentVisitorContext> base_context) {
-  auto context = std::static_pointer_cast<Context>(base_context);
-  BaseSingleColumnTableScanImpl::handle_segment(base_segment, base_context);
+std::shared_ptr<PosList> IsNullTableScanImpl::scan_chunk(const ChunkID chunk_id) const {
+  const auto& chunk = _in_table->get_chunk(chunk_id);
+  const auto& segment = chunk->get_segment(_column_id);
 
-  const auto& pos_list = *base_segment.pos_list();
+  auto matches = std::make_shared<PosList>();
 
-  // Additionally to the null values in the referencED segment, we need to find null values in the referencING segment
-  if (_predicate_condition == PredicateCondition::IsNull) {
-    for (ChunkOffset chunk_offset{0}; chunk_offset < pos_list.size(); ++chunk_offset) {
-      if (pos_list[chunk_offset].is_null()) context->_matches_out.emplace_back(context->_chunk_id, chunk_offset);
-    }
-  }
+  _on_scan(*segment, chunk_id, *matches, nullptr);
+
+  return matches;
+}
+
+void IsNullTableScanImpl::_on_scan(const BaseSegment& segment, const ChunkID chunk_id, PosList& matches,
+                                   const std::shared_ptr<const PosList>& position_filter) const {
+  resolve_data_and_segment_type(segment, [&](const auto type, const auto& typed_segment) {
+    _scan_segment(typed_segment, chunk_id, matches, position_filter);
+  });
 }
 
-void IsNullTableScanImpl::handle_segment(const BaseValueSegment& base_segment,
-                                         std::shared_ptr<SegmentVisitorContext> base_context) {
-  auto context = std::static_pointer_cast<Context>(base_context);
-  const auto& position_filter = context->_position_filter;
+void IsNullTableScanImpl::_scan_segment(const BaseSegment& segment, const ChunkID chunk_id, PosList& matches,
```

Comment:
(Random place): Maybe calling these `_scan_with_resolved_segment_type` (or similar) makes it clearer why we need to have `on_scan`, `_on_scan` and then **also** `_scan_segment`?

Reply:
I get what you mean, but as we also have the general case `BaseSegment&` (which you attached this comment to), I have some reservations with calling it "resolved".

Where do we have `on_scan`?

Reply:
ping?

Reply:
Apparently my answer was lost...

I didn't mean on_scan, I meant "scan_chunk". 

"scan_chunk()" as a name is okay, it does what the name implies
"_on_scan()" could be named better, maybe "scan_non_reference_chunk/segment"?
"_scan_segment()" I suppose I will have to live with
"_scan_with_iterators" seems fine as well.

Reply:
done